### PR TITLE
Enable hash-checking mode for pip install

### LINKED ISSUE #137: Enable hash-checking mode for pip install?
Hi!

**Description of issue or feature request**:

It might be a good idea to [add hashes](https://pip.pypa.io/en/stable/reference/pip_install/#hash-checking-mode) to packages installed via ```pip```, such as ```certifi``` and ```auditwheel```.

Upside: This should allow a build of ```manylinux``` which uses ```pip``` to detect whether or not packages have been modified against expectations.

Downside: This requires manually updating dependencies (but that might be acceptable in this case, as we do pin down the versions and hashes of many other packages).

**Current behavior**:

No hashes are listed.

**Expected behavior**:

Hashes should be listed and checked.

**Miscellaneous notes**:

I'll send a PR when I find some downtime.

### PR DESCRIPTION
**Pull request by trishankatdatadog**
It might be a good idea to [add hashes](https://pip.pypa.io/en/stable/reference/pip_install/#hash-checking-mode) to packages installed via ```pip```, such as ```certifi``` and ```auditwheel```.

Upside: This should allow a build of ```manylinux``` which uses ```pip``` to detect whether or not packages have been modified against expectations.

Downside: This requires manually updating dependencies (but that might be acceptable in this case, as we do pin down the versions and hashes of many other packages).

This fixes #137.

### CODE REVIEW

**Comment by njsmith** [comment]
I'm nervous about the security/maintainability trade-off here.

The upside of this is that it gives us some protection against a compromise of PyPI (but only if the attackers modify old releases instead of making new ones), or a MITM attack on our connection to PyPI. But the latter shouldn't happen, since by the time we're running pip we're supposed to have working TLS etc. Plus, the main (only?) reason someone would want to compromise the manylinux docker image is so that they could compromise packages when they're built before they're uploaded to PyPI, so if someone's already pwned PyPI then we've already lost.

The downside is that it substantially increases the maintenance burden for the docker image, because now we have to manually update these hashes every time any of these packages make a release, which is something that happens pretty often. In practice this probably means that we just won't update these packages most of the time, which means that we won't be getting any security fixes that they release. This seems like a bigger problem than the one we were originally trying to solve.

What do you think? Is there some threat model I'm not thinking of?

**Comment by njsmith** [comment]
Note that we currently don't have any kind of cryptographic checks on our actual compilers that we're getting from http://people.centos.org/tru/devtools-2/devtools-2.repo – no TLS, no package signatures, nothing. This is probably our weakest link (aside from attacks on developer laptops with commit access). It does look like people.centos.org supports HTTPS now (but the .repo file there doesn't use it); maybe using our own copy of that repo file with a https:// URL swapped in would be a good way to improve the security there?

**Comment by trishankatdatadog** [comment]
On the upside: I don't think you need to compromise all of PyPI. The attacker just needs to, e.g., steal credentials for one of the few (5-6?) PyPI packages used in manylinux. So, in that case, we would be protected.

On the downside: you are right, it does increase the maintenance burden of updating requirements.txt, but only when there is a _security_ update for one of these PyPI packages. I think that has been always been the case anyway for other software included in manylinux (e.g., openssl had to be updated just a weeks ago). To alleviate the burden, there are some automatic tools (e.g., [pyup](https://pyup.io/)) that can be used to monitor whether pinned dependencies now suffer from known vulnerabilities.

To summarize, I agree with you that there is increased maintenance burden, but I think the need to update software on manylinux for security fixes has always been there anyway. For example, even if you don't fix the PyPI package hashes as in here, you will probably need to build and reupload new Docker images anyway when security updates to these packages arise (because the old images would contain the vulnerabilities these updates address).

I hope this makes sense. What do you, and the other maintainers, think?

**Comment by trishankatdatadog** [comment]
Regarding the devtools-2 repo: it looks like those packages have not been updated in three years. It might just be simplest to bundle a static copy of those packages with this GitHub repository, if possible.

**Comment by njsmith** [comment]
> I think that has been always been the case anyway for other software included in manylinux (e.g., openssl had to be updated just a weeks ago).

Indeed, but right now openssl is the main thing that needs to be updated manually; adding another 7 items to that list is a big change, especially since projects like setuptools release much more frequently than openssl does.

Setting up something like pyup or requires.io would indeed make sense to improve the promptness of our rebuilds, independently of this PR. Do they handle updating the hashes?

**Comment by trishankatdatadog** [comment]
I totally understand, @njsmith, and will leave this decision to the maintainers.

Looking at its [config](https://pyup.io/docs/bot/config/), I believe pyup can update requirements.txt files, but this needs testing.

I know TUF is using pyup. @vladimir-v-diaz, sorry to bother you, but would you happen to know if pyup can automatically update your requirements.txt files for you?

**Comment by vladimir-v-diaz** [comment]
@trishankatdatadog 
Yes.  It also [supports pip hashes](https://github.com/pyupio/pyup/commit/3bf5ac9232f82f7e5bf2db0cf243e2f7402643b8).

**Comment by trishankatdatadog** [comment]
Thanks, @vladimir-v-diaz, good to know!

**Comment by njsmith** [comment]
Okay, I just enabled pyup.io, which I *think* should address all my concerns... I guess we can see how it goes.

**Comment by trishankatdatadog** [comment]
Thank you very much, sir! 🙏  @njsmith
